src/main/java/com/slack/kaldb/server/KaldbQueryService.java

```diff
@@ -15,26 +25,70 @@
 
   public static List<String> servers = new ArrayList<>();
 
-  /**
-   * TODO 1. Cache the stub 2. Run in parallel 3. Integrate with ZK to update list 4. Aggregate the
-   * responses instead of only returning the result from the first server
-   */
-  public KaldbSearch.SearchResult distributedSearch(KaldbSearch.SearchRequest request) {
+  // public so that we can override in tests
+  // TODO: In the future expose this as a config in the proto
+  // TODO: ChunkManager#QUERY_TIMEOUT_SECONDS and this could be unified?
+  public static int READ_TIMEOUT_MS = 15000;
+
+  public static final KaldbSearch.SearchResult error =
+      KaldbSearch.SearchResult.newBuilder().setFailedNodes(1).setTotalNodes(0).build();
+
+  // TODO Cache the stub
+  // TODO Integrate with ZK to update list of servers
+
+  public CompletableFuture<List<KaldbSearch.SearchResult>> distributedSearch(
+      KaldbSearch.SearchRequest request) {
+
+    List<CompletableFuture<KaldbSearch.SearchResult>> futures = new ArrayList<>(servers.size());
+
     for (String server : servers) {
-      KaldbServiceGrpc.KaldbServiceBlockingStub stub =
-          Clients.newClient(server, KaldbServiceGrpc.KaldbServiceBlockingStub.class);
-      return stub.search(request);
+      // With the deadline we are keeping a high limit on how much time can each CompletableFuture
+      // take
+      // Alternately use completeOnTimeout and the value can then we configured per request and not
+      // as part of the config
+      KaldbServiceGrpc.KaldbServiceFutureStub stub =
+          Clients.newClient(server, KaldbServiceGrpc.KaldbServiceFutureStub.class)
+              .withDeadlineAfter(READ_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+
+      futures.add(ListenableFuturesExtra.toCompletableFuture(stub.search(request)));
     }
-    throw new RuntimeException("no live servers");
+    return futures
+        .stream()
+        .map(result -> result.exceptionally(ex -> error))
+        .collect(CompletableFutures.joinList());
   }
 
   @Override
   public void search(
       KaldbSearch.SearchRequest request,
       StreamObserver<KaldbSearch.SearchResult> responseObserver) {
 
-    KaldbSearch.SearchResult protoSearchResult = distributedSearch(request);
-    responseObserver.onNext(protoSearchResult);
-    responseObserver.onCompleted();
+    CompletableFuture<List<SearchResult<LogMessage>>> searchResults =
```

Comment:
Please add function doc explaining the logic in search and distributedSearch functionality. Why do we need to call fromSearchResultAndIncrementNodeCount? Isn't the local searcher already incrementing the node count?

Reply:
done!